src/Type/Doctrine/Query/QueryAggregateFunctionDetectorTreeWalker.php

```diff
@@ -0,0 +1,321 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Doctrine\Query;
+
+use Doctrine\ORM\Query;
+use Doctrine\ORM\Query\AST;
+use function is_string;
+
+class QueryAggregateFunctionDetectorTreeWalker extends Query\TreeWalkerAdapter
```

Comment:
The logic is copied from Doctrine's SqlWalker. But since they do double-dispatch which is not possible here, those places are mostly replaced with generic doWalkNode.

I also prefixed all methods with `do` so that this works in dbal3 and dbal4 (where TreeWalker interface lost most of its methods).